src/util/BallFilter.cpp

```diff
@@ -0,0 +1,138 @@
+    startCov.eye();
+    //initial noise estimates
+    const double startPosNoise = 0.3;
+    const double startAngleNoise = 0.05;
+    startCov.at(0, 0) = startPosNoise;//m noise in x
+    startCov.at(1, 1) = startPosNoise;//m noise in y
+
+    kalman = std::make_unique<Kalman>(startState, startCov);
+
+    kalman->H.eye();     // Our observations are simply what we see.
+
+    const double posVar = 0.02; //variance in meters
+    kalman->R.zeros();
+    kalman->R.at(0, 0) = posVar;
+    kalman->R.at(1, 1) = posVar;
+}
+void BallFilter::applyObservation(const proto::SSL_DetectionBall &detectionBall) {
+    Kalman::VectorO observation;
+    observation.zeros();
+    observation.at(0) = mmToM(detectionBall.x());
+    observation.at(1) = mmToM(detectionBall.y());
+
+    //TODO: do things with the other ball fields (pixel pos, area)
+    kalman->z = observation;
+    kalman->update();
+}
+int BallFilter::frames() const {
+    return frameCount;
+}
+double BallFilter::getLastFrameTime() const {
+    return lastUpdateTime;
+}
+proto::WorldBall BallFilter::asWorldBall() const {
+    proto::WorldBall msg;
+    const Kalman::Vector &state = kalman->state();
+    msg.mutable_pos()->set_x(state[0]);
+    msg.mutable_pos()->set_y(state[1]);
+    msg.mutable_vel()->set_x(state[2]);
+    msg.mutable_vel()->set_y(state[3]);
+    msg.set_visible(ballIsVisible());
+    //TODO: add height filter here, and actually set the z, z_vel, area fields
+    return msg;
+}
+double BallFilter::distanceTo(double x, double y) const {
+    const Kalman::Vector &state = kalman->state();
+    double dx = state[0] - mmToM(x);
+    double dy = state[1] - mmToM(y);
+    return sqrt(dx * dx + dy * dy);
+}
+void BallFilter::predict(double time, bool permanentUpdate) {
+    double dt = time - lastUpdateTime;
+    // forward model:
+    kalman->F.eye();
+    kalman->F.at(0, 2) = dt;
+    kalman->F.at(1, 3) = dt;
+
+    //TODO: add 2 stage forward model?
+    //Set B
+    kalman->B = kalman->F;
+    //Set u (we have no control input at the moment)
+    kalman->u.zeros();
+
+    //Set Q matrix
+    Kalman::MatrixO G;
+    G.zeros();
+    G.at(0, 0) = dt;
+    G.at(0, 2) = 1;
+    G.at(1, 1) = dt;
+    G.at(1, 3) = 1;
+
+    const float processNoise = 0.01;
+    kalman->Q = G.t() * G * processNoise;
+
+    kalman->predict(permanentUpdate);
+    lastPredictTime=time;
+    if (permanentUpdate) {
+        lastUpdateTime = time;
+    }
+}
+bool compareObservation(const BallFilter::BallObservation& a, const BallFilter::BallObservation& b ){
+    return (a.time<b.time);
+}
+void BallFilter::update(double time, bool doLastPredict) {
+
+    std::sort(observations.begin(),observations.end(),compareObservation); //First sort the observations in time increasing order
+    auto it=observations.begin();
+    while(it != observations.end()) {
+        auto observation=(*it);
+        //the observation is either too old (we already updated the robot) or too new and we don't need it yet.
```

Comment:
I don't think you are understanding it fully correctly. The observation sorted data ques indeed. In order for kalman filters to work properly we need to do predict+applyObservation for the right time (see line 122, 123) and step forwards in time through all our observations. 
Timing is difficult to explain properly (I kind of need a whiteboard), but the way it works is like this: Our program ticks independently of when vision packets arrive, and everytime we tick we do update-predict cycles for the kalman-filters until the latest known time of all the frames. (This is also why we sort the frames by time first). The 'time' parameter we pass this BallFilter::update function is typically larger or equal than the latest time of the frame. If doLastPredict is set to true, the filter will also do extrapolation from the last time we saw a robot to the current time requested so that all things in the world can be seen at the same point in time. I wrote this function without any real-time requirements so that we can also use it for analysis and that the behavior is deterministic for the same data sets regardless of the timing intervals of the loop this is ran in.

Reply:
I don't think you are understanding it fully correctly. The observation sorted data ques indeed. In order for kalman filters to work properly we need to do predict+applyObservation for the right time (see line 122, 123) and step forwards in time through all our observations.
Timing is difficult to explain properly (I kind of need a whiteboard), but the way it works is like this: Our program ticks independently of when vision packets arrive, and everytime we tick we do update-predict cycles for the kalman-filters until the latest known time of all the frames. (This is also why we sort the frames by time first). The 'time' parameter we pass this BallFilter::update function is typically larger or equal than the latest time of the frame. If doLastPredict is set to true, the filter will also do extrapolation from the last time we saw a robot to the current time requested so that all things in the world can be seen at the same point in time. I wrote this function without any real-time requirements so that we can also use it for analysis and that the behavior is deterministic for the same data sets regardless of the timing intervals of the loop this is ran in.